doc/md/fundamentals/3-declarations/4-class-declarations.md

```diff
@@ -0,0 +1,141 @@
+
+Use a class when you need to:
+- Create multiple objects with the same fields and methods.
+- Encapsulate state and behavior inside an object.
+- Simulate object orientated programming-like design patterns.
+
+Classes are particularly useful when using multiple objects with similar structures but different initial states. You can think of classes as a template for creating objects. You define fields and methods, and then create new instances by calling the class like a function.
+ 
+## Defining a class
+
+A class in Motoko defines both:
+
+- A constructor for creating instances, which includes:
+  - Fields to store data.
+  - Methods to perform operations on that data.
+- A type that describes the interface of the instance, including its available methods.
+
+Motoko also supports generic classes, which allow abstraction over types. This enables the creation of reusable, type-safe components that can work with a variety of data types while maintaining strong compile-time guarantees.
+
+```motoko no-repl
+import O "mo:base/Order"; // Required to reference O.Order and its variants
+
+  // A generic comparator class for ordering elements
+  class Comparator<T>(compare : (T, T) -> O.Order) {
+    public func lessThan(a : T, b : T) : Bool {
+      compare(a, b) == #less;
+    };
+
+    public func equal(a : T, b : T) : Bool {
+      compare(a, b) == #equal;
+    };
+
+    public func greaterThan(a : T, b : T) : Bool {
+      compare(a, b) == #greater;
+    };
+  };
+
+```
+
+This declaration creates:
+
+- A constructor: `Comparator : ((K, K) -> Order) -> Comparator<K>`.
+- A generic type: `Comparator<K>` that provides comparison methods.
+
+## Creating instances
+
+Creating an instance is when a class constructor is used to produce a new object with its own state and behavior based on the class definition.
+
+```motoko no-repl
+class Counter(init : Nat) {
+  var count : Nat = init;
+
+  public func inc() : Nat {
+    count += 1;
+    count
+  };
+};
+
+// Creating an instance of the class
+let myCounter = Counter(0);
+
+// Using the instance
+myCounter.inc(); // returns 1
+```
+
+## Actor classes
+
+Actor classes enable you to create networks of actors programmatically. These classes must be defined in separate source files.
+
+```motoko no-repl title="Counter.mo"
+actor class Counter() {
+  var count : Int = 0;
+
+  public func increment() : async Int {
+    count += 1;
+    return count;
+  };
+
+  public func get() : async Int {
+    return count;
+  };
+}
+```
+
+This defines a `Counter` actor class with:
+
+- A mutable variable `count`.
+
+- A method `increment` that increases `count` and returns the new value.
+
+- A method get that returns the current count.
+
+Then, in another file you can create an instance of this actor class:
+
+```motoko no-repl title="CallCounter.mo"
+import Counter "./Counter";
+
+actor {
+  let counter = await (with cycles = 1_000_000_000_000) Counter.Counter();
```

Comment:
This shouldn't typecheck since you can't await or sent messages in the body of an actor.

See the suggested rewrite below.